src/thor/bidirectional_astar.cc

```diff
@@ -1053,102 +1069,67 @@ std::vector<std::vector<PathInfo>> BidirectionalAStar::FormPath(GraphReader& gra
-            path.back().elapsed_cost.cost = std::max(cost.cost, 0.f);
-            not_found = false;
-          }
-        }
-
-        // This cannot happen because we only make labels from edge candidates in the destination
-        // location
-        if (not_found)
-          throw std::logic_error("Could not find candidate edge used for destination label");
-      }
-    }
-
-    // Get the elapsed time at the end of the forward path. NOTE: PathInfo
-    // stores elapsed time as uint32_t but EdgeLabels uses float. Need to
-    // accumulate in float and cast to int so we do not accumulate roundoff
-    // error.
-    Cost cost = path.back().elapsed_cost;
-
-    // Get the transition cost at the last edge of the reverse path
-    Cost previous_transition_cost = edgelabels_reverse_[idx2].transition_cost();
+    std::reverse(path_edges.begin(), path_edges.end());
 
     // Append the reverse path from the destination - use opposing edges
     // The first edge on the reverse path is the same as the last on the forward
     // path, so get the predecessor.
-    uint32_t edgelabel_index = edgelabels_reverse_[idx2].predecessor();
-    while (edgelabel_index != kInvalidLabel) {
+    for (auto edgelabel_index = edgelabels_reverse_[idx2].predecessor();
+         edgelabel_index != kInvalidLabel;
+         edgelabel_index = edgelabels_reverse_[edgelabel_index].predecessor()) {
       const BDEdgeLabel& edgelabel = edgelabels_reverse_[edgelabel_index];
-      GraphId oppedge = graphreader.GetOpposingEdgeId(edgelabel.edgeid());
-
-      // The first edge has no previous edge with elapsed cost so we can just use it directly
-      uint32_t predidx = edgelabel.predecessor();
-      if (predidx == kInvalidLabel) {
-        cost += edgelabel.cost();
-      } // This edge needs the elapsed cost between it and the previous edge, since its the reverse
-        // path in flipping the path around we also need to switch out the transition costs (shifting
-        // right)
-      else {
-        cost += edgelabel.cost() - edgelabels_reverse_[predidx].cost() - edgelabel.transition_cost();
-      }
-      cost += previous_transition_cost;
-      path.emplace_back(edgelabel.mode(), cost, oppedge, 0, edgelabel.restriction_idx(),
-                        previous_transition_cost);
+      const DirectedEdge* opp_edge = nullptr;
+      GraphId opp_edge_id = graphreader.GetOpposingEdgeId(edgelabel.edgeid(), opp_edge, tile);
+
+      if (opp_edge->is_shortcut()) {
+        auto superseded = graphreader.RecoverShortcut(opp_edge_id);
+        std::move(superseded.begin(), superseded.end(), std::back_inserter(path_edges));
+      } else
+        path_edges.emplace_back(std::move(opp_edge_id));
 
       // Check if this is a ferry
       if (edgelabel.use() == Use::kFerry) {
         has_ferry_ = true;
       }
+    }
+
+    // don't care about trivial paths here, it will be handled correctly in recost function below
```

Comment:
the comment is wrong here. there is an existing bug with bidirectional a* where routes that start and end on the same edge fail when they need to go around the block.

```suggestion
    // bidirectional a* has a bug where it fails trivial routes in which you are on a one way edge and
    // the origin is near the end of the edge and the destination is near the beginning, in other words
    // a route that looks trivial but actually needs to go around the block to complete
```